### PR TITLE
newstore:  removing deprecated files when use new store driver

### LINKED ISSUE #803: [RFC] baseline persist data for live-upgrade
As part of discussion: https://github.com/kata-containers/runtime/issues/492

>1) kata-runtime
>A. issue "versioned" command to kata-agent, can always communicate correctly with old kata-agent. (MUST)
>B. disk persist data should be "versioned", kata-runtime can always handle old "version" of persist data to restore sandbox/container struct from disk to memory. (MUST)

The disk persist data should be "versioned" or baselined, and has a good rule for future modification. Currently we modify the data quite casually which is bad.

The persist data can include data under "/run/vc/sbs", "/var/lib/vc/sbs/" and "/run/kata-containers/",  among them the "/run/vc/sbs" should be most important and difficult part.

These files including:
```
/run/vc/sbs/d827f23e7157863f23924f9d54d3822cd96f932f9ce53935b6fc7a1e61634b01
├── agent.json
├── d827f23e7157863f23924f9d54d3822cd96f932f9ce53935b6fc7a1e61634b01
│   ├── devices.json
│   ├── mounts.json
│   ├── process.json
│   └── state.json
├── devices.json
├── hypervisor.json
├── lock
├── network.json
├── proxy.sock
└── state.json
```

Based on this description, I have this proposal:

## **1. combine these files to one file**

Currently we have many files here, and I think `agent.json`, `devices.json`, `hypervisor.json`, `network.json` and `state.json` can be combined to one. I don't think seperating them into several files can bring many benefits, this may bring some discussions or argues about performance penalty caused by disk writting, let's discuss whether this is real necessary.

## **2.  baseline the persist data format.**

Let me take https://github.com/kata-containers/runtime/pull/32 as an example, persist data format should be regarded as one kind of exported APIs. 

We can create a dir `virtcontainers/persist/types.go` containing persist data format:

e.g.

```
// DeviceInfo is an embedded type that contains device data common to all types of devices.
type DeviceInfo struct {
    // Hostpath is device path on host
    HostPath string

    // ContainerPath is device path inside container
    ContainerPath string

    // Type of device: c, b, u or p
    // c , u - character(unbuffered)
    // p - FIFO
    // b - block(buffered) special file
    // More info in mknod(1).
    DevType string

    // Major, minor numbers for device.
    Major int64
    Minor int64

    // FileMode permission bits for the device.
    FileMode os.FileMode

    // id of the device owner.
    UID uint32

    // id of the device group.
    GID uint32

    // Hotplugged is used to store device state indicating if the
    // device was hotplugged.
    Hotplugged bool

    // ID for the device that is passed to the hypervisor.
    ID string

    // DriverOptions is specific options for each device driver
    // for example, for BlockDevice, we can set DriverOptions["blockDriver"]="virtio-blk"
    DriverOptions map[string]string
}
```

We should define all the data we want to export in a consolidated dir, and keeps an eye on any change to this dir.

And we can have a `Dump()` and `Restore()` function to convert internal structure to exported data structure and vise versa,

`RunV` gives a good example: https://github.com/hyperhq/runv/blob/master/hypervisor/persistence.go

### PR DESCRIPTION
**Pull request by WeiZhang555**
Fixes #803

It contains several commits:

1. Simplify new store API to make the code easier to understand and use.
2. newstore: remove file "devices.json" 

Signed-off-by: Wei Zhang <zhangwei555@huawei.com>

### CODE REVIEW

**Comment by WeiZhang555** [comment]
Second commit is only for test purpose which should be removed before merging.
After we enabled experimental Jenkins job, we don't need the test commit any more.

ping CI team @kata-containers/ci : 

https://github.com/kata-containers/runtime/pull/883#issuecomment-485056458

File: virtcontainers/api.go [hunk 1]
```diff
@@ -590,6 +590,10 @@ func statusContainer(sandbox *Sandbox, containerID string) (ContainerStatus, err
 				}
 
 				if !running {
+					virtLog.WithFields(logrus.Fields{
+						"state": container.state.State,
+						"Pid":   container.process.Pid}).
```
Comment (jodh-intel) [inline virtcontainers/api.go]: Nit: I'd make the field name lowercase: `"pid"`.

File: virtcontainers/device/drivers/block.go [hunk 2]
```diff
@@ -169,5 +169,26 @@ func (device *BlockDevice) Dump() persistapi.DeviceState {
 	return ds
 }
 
+// Load loads DeviceState and converts it to specific device
+func (device *BlockDevice) Load(ds persistapi.DeviceState) {
+	device.GenericDevice = &GenericDevice{}
+	device.GenericDevice.Load(ds)
+
+	bd := ds.BlockDrive
+	if bd != nil {
```
Comment (jodh-intel) [inline virtcontainers/device/drivers/block.go]: Nit: I tend to invert logic like this to reduce the indent and make the code as simple as possible:

```go

if bd == nil {
    return
}

device.BlockDrive = &config.BlockDrive{
      :
```

Same comment for some of the checks below...

File: virtcontainers/device/manager/manager.go [hunk 3]
```diff
@@ -229,3 +230,39 @@ func (dm *deviceManager) IsDeviceAttached(id string) bool {
...
+			dm.devices[dev.DeviceID()] = dev
+		case config.VhostUserBlk:
+			dev := &drivers.VhostUserBlkDevice{}
+			dev.Load(ds)
+			dm.devices[dev.DeviceID()] = dev
+		case config.VhostUserNet:
+			dev := &drivers.VhostUserNetDevice{}
+			dev.Load(ds)
+			dm.devices[dev.DeviceID()] = dev
+		}
```
Comment (jodh-intel) [inline virtcontainers/device/manager/manager.go]: How about adding a `default` case and logging a warning/error that the type cannot be handled?

File: netmon/netmon.go [hunk 4]
```diff
@@ -567,26 +567,26 @@ func (n *netmon) handleRTMDelRoute(ev netlink.RouteUpdate) error {
 }
 
 func (n *netmon) handleLinkEvent(ev netlink.LinkUpdate) error {
-	n.logger().Debug("handleLinkEvent: netlink event received")
+	n.logger().Info("handleLinkEvent: netlink event received")
 
 	switch ev.Header.Type {
 	case unix.NLMSG_DONE:
-		n.logger().Debug("NLMSG_DONE")
+		n.logger().Info("NLMSG_DONE")
```
Comment (jodh-intel) [inline netmon/netmon.go]: How many of these events do we normally get? I'm wondering if `Info` might end up DoS'ing the logs so best to leave at `Debug` level?

File: virtcontainers/persist.go [hunk 5]
```diff
@@ -61,7 +67,7 @@ func deviceToDeviceState(devices []api.Device) (dss []persistapi.DeviceState) {
 	return
```
Comment (gabibeyer) [inline virtcontainers/persist.go]: Although not really part of this PR, you could probably get rid of this return if you feel like it goes with some of the other clean up.

File: netmon/netmon.go (hunk 4 above)
Comment (WeiZhang555) [reply netmon/netmon.go]: Yeah, this is only for debug purpose, I ran into some wired problems in CI and I don't have a good environment to reproduce, still struggling...

These changes will be removed in final version. Thanks :-)

File: virtcontainers/device/drivers/block.go (hunk 2 above)
Comment (WeiZhang555) [reply virtcontainers/device/drivers/block.go]: Oh, nice suggestion!

File: virtcontainers/persist.go (hunk 5 above)
Comment (WeiZhang555) [reply virtcontainers/persist.go]: Right 👍

File: virtcontainers/api.go (hunk 1 above)
Comment (WeiZhang555) [reply virtcontainers/api.go]: done

**Comment by WeiZhang555** [comment]
ping @chavafg  : Can you enable "experimental" in `cri-containerd` job so that we can test experimental features with K8s test cases? I think cri-containerd+Fedora can cover almost all cases for experimental features.

I can't find how to set the "KATA_EXPERIMENTAL_FEATURES" flag in our Jenkins Jobs, maybe that's because I don't have right permission.

**Comment by WeiZhang555** [comment]
/test

@jodh-intel @gabibeyer  I fixed the CI now, so please take another look at this. Thanks :smile:

**Comment by chavafg** [comment]
Hi @WeiZhang555, I've just enabled it in the cri-containerd job and also provided you with correct permissions on Jenkins.

**Comment by WeiZhang555** [comment]
Thanks so much! @chavafg

**Comment by WeiZhang555** [comment]
ping @kata-containers/runtime

File: virtcontainers/device/manager/manager.go [hunk 6]
```diff
@@ -229,3 +230,41 @@ func (dm *deviceManager) IsDeviceAttached(id string) bool {
...
+
+	for _, ds := range devStates {
+		switch config.DeviceType(ds.Type) {
+		case config.DeviceGeneric:
+			dev := &drivers.GenericDevice{}
+			dev.Load(ds)
+			dm.devices[dev.DeviceID()] = dev
+		case config.DeviceBlock:
+			dev := &drivers.BlockDevice{}
+			dev.Load(ds)
```
Comment (jodh-intel) [inline virtcontainers/device/manager/manager.go]: Nit: It would be great if we could avoid all the `dev.Load()` and `dm.devices[dev.DeviceID()]` duplication somehow  (an interface?) so that those calls could be made once, after the `switch`

File: virtcontainers/persist.go [hunk 7]
```diff
@@ -90,59 +96,53 @@ func (s *Sandbox) dumpDevices(ss *persistapi.SandboxState, cs map[string]persist
...
+	return nil
 }
 
-func (s *Sandbox) getSbxAndCntStates() (*persistapi.SandboxState, map[string]persistapi.ContainerState, error) {
-	if err := s.newStore.FromDisk(s.id); err != nil {
-		return nil, nil, err
-	}
+func (s *Sandbox) loadState(ss persistapi.SandboxState) {
+	s.state.PersistVersion = ss.PersistVersion
+	s.state.GuestMemoryBlockSizeMB = ss.GuestMemoryBlockSizeMB
```
Comment (jodh-intel) [inline virtcontainers/persist.go]: The only element of `SandboxState` you are not setting here is `GuestMemoryHotplugProbe`. Is there any reason for that? Might be worth a comment explaining why if so.

File: virtcontainers/api.go [hunk 8]
```diff
@@ -590,6 +590,10 @@ func statusContainer(sandbox *Sandbox, containerID string) (ContainerStatus, err
 				}
 
 				if !running {
+					virtLog.WithFields(logrus.Fields{
+						"state":       container.state.State,
+						"process pid": container.process.Pid}).
```
Comment (jodh-intel) [inline virtcontainers/api.go]: Nit: By (admittedly undocumented) convention we try to ensure all log field names do not contain spaces, so could you change this to `process-id` (or even just `pid` as I think that's a well-understood term).

File: virtcontainers/persist.go (hunk 7 above)
Comment (WeiZhang555) [reply virtcontainers/persist.go]: Nice catch! This is a new field recently added, I will add save/load logic here.

File: virtcontainers/device/manager/manager.go (hunk 6 above)
Comment (WeiZhang555) [reply virtcontainers/device/manager/manager.go]: Sounds good.

File: virtcontainers/api.go (hunk 8 above)
Comment (WeiZhang555) [reply virtcontainers/api.go]: I'll use the `process-pid` since we already have a `pid` field in `virtLog`, so if I use `pid` here that will confuse developers.

**Comment by WeiZhang555** [comment]
/test

Thanks @jodh-intel ! Your comments are addressed.

File: virtcontainers/persist.go (hunk 7 above)
Comment (jodh-intel) [reply virtcontainers/persist.go]: Thanks @WeiZhang555.

Thinking about this, I wonder if there is a way we could use reflection to ensure we always copy all elements in all structs? I think it would be possible as you could iterate through `ss` and then match the corresponding element in `s` based on name and type. It would complicate the code though.

Alternatively, maybe you could create a unit test that:

- Sets `ss` so that all elements have non-zero/non-`nil` values.
- Call `s.loadState()` with a new `Sandbox`.
- Iterate over all the fields in `s.state` and ensure they are all set, and set to the values in `ss`.

**Comment by gabibeyer** [comment]
lgtm!

**Comment by teawater** [comment]
jenkins-ci-centos-7-4-q-35 fail:
```
--- FAIL: TestHostNetworkingRequested (0.01s)
    assertions.go:239: 
                          
	Error Trace:	network_test.go:99
        
	Error:      	Should be true
```
This is a known issue https://github.com/kata-containers/runtime/issues/1604

**Comment by WeiZhang555** [comment]
Thanks @teawater !
